### ISSUE TITLE
Fails with error with "mark" and "remove"

### ISSUE THREAD

**Issue by CarlosNZ** [mxmvshnvsk/i18n-unused#16]
It's working fine (and giving seemingly correct results) with "display-unused". However, when I run "mark-unused" or "remove-unused", I get an error like so:

```
 Successfully marked: /Users/carl/GitHub/opensupply/client/packages/common/src/intl/locales/ar/app.json
(node:18673) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'admin' of undefined
    at /Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:469:51
    at /Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:384:7
    at Array.reduce (<anonymous>)
    at applyToFlatKey (/Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:382:16)
    at /Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:468:37
    at Array.forEach (<anonymous>)
    at /Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:468:22
    at Array.forEach (<anonymous>)
    at markUnusedTranslations (/Users/carl/GitHub/opensupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:465:35)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:18673) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:18673) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

It deals with the first file (`app.json`) fine, but then gets no further.

Any ideas?

**Comment by mxmvshnvsk** [comment]
Hi. Can u attach part of `app.json` and your `i18n-unused.config.js`? it hard to  say, where the problem by stack trace.

**Comment by CarlosNZ** [comment]
[i18n-unused-test.zip](https://github.com/mxmvshnvsk/i18n-unused/files/8290902/i18n-unused-test.zip)

Thanks. This .zip contains:
- `i18n-unused.config.js`
- `app.json` (which seems to be the only file that is behaving correctly)
- `common.json` (which I think is the next file analysed which throws the error)

You can clone our repo if you'd like to see more detail: https://github.com/openmsupply/openmsupply-client

Many thanks. :)

**Comment by mxmvshnvsk** [comment]
Ok, I'll take a look this week.

**Comment by mxmvshnvsk** [comment]
I looked at your code. U have errors, because u use wrong deep translations. For example in [`distribution.json`](https://github.com/openmsupply/openmsupply-client/blob/3e88ff377f54b47d41e9306cbf91585c9e22bcab/packages/common/src/intl/locales/en/distribution.json#L42) u use `"label.group-by-item": "Group by Item",` instead of [object structure](https://www.i18next.com/translation-function/essentials#accessing-keys) for [`t('label.group-by-item')`](https://github.com/openmsupply/openmsupply-client/blob/3e88ff377f54b47d41e9306cbf91585c9e22bcab/packages/invoices/src/OutboundShipment/DetailView/ContentArea.tsx#L114). So, when my lib builds the structure `t('label.group-by-item')` back, it expands into an object, not a string, e.g.:

```javascript
// wrong
{
  ...
  "label.group-by-item": "Group by Item",
  ...
}

// change to
{
  ...
  "label": {
    "group-by-item": "Group by Item"
  }
  ...
}
```

If I helped solve your problem, please, close an issue. Also will be glad for star =)

**Comment by mxmvshnvsk** [comment]
@CarlosNZ, have u any updates?

**Comment by mxmvshnvsk** [comment]
@CarlosNZ, do you have any updates?

**Comment by CarlosNZ** [comment]
Okay thanks, I'll take a closer look tomorrow and close the issue.

**Comment by CarlosNZ** [comment]
Hi, it seems like "flat" nested keys should be acceptable in i18n JSON files: https://www.codeandweb.com/babeledit/documentation/file-formats#flat-json

And indeed we are using `t('label.group-by-item')`-style using the https://www.i18next.com/ package throughout our app with no problems.

**Comment by mxmvshnvsk** [comment]
Ok, I see your point. I did a little research on this topic. I think it's possible to make implementation for flat json in next minor version at this week.

**Comment by CarlosNZ** [comment]
Excellent, thank you :)

**Comment by mxmvshnvsk** [comment]
Hi. Thx for waiting =) U can try new [0.8.0](https://www.npmjs.com/package/i18n-unused/v/0.8.0) version. Now available option `flatTranslations `, boolean, switch locale builder to flat json mode.

**Comment by CarlosNZ** [comment]
Thanks @mxmvshnvsk ,

it's working better now (gets further), but it's still throwing an error at a certain point when running `mark`:

```
Successfully marked: /Users/carl/GitHub/openmsupply/client/packages/common/src/intl/locales/ar/app.json
 Successfully marked: /Users/carl/GitHub/openmsupply/client/packages/common/src/intl/locales/ar/common.json
 Successfully marked: /Users/carl/GitHub/openmsupply/client/packages/common/src/intl/locales/ar/replenishment.json
 Successfully marked: /Users/carl/GitHub/openmsupply/client/packages/common/src/intl/locales/en/app.json
(node:63505) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'default-sell-price' of undefined
    at applyToFlatKey.flatTranslations (/Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:485:51)
    at /Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:395:7
    at Array.reduce (<anonymous>)
    at applyToFlatKey (/Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:393:16)
    at /Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:484:37
    at Array.forEach (<anonymous>)
    at /Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:484:22
    at Array.forEach (<anonymous>)
    at markUnusedTranslations (/Users/carl/GitHub/openmsupply/client/node_modules/i18n-unused/dist/i18n-unused.cjs:481:35)
```
Oddly, there doesn't seem to be anything unique about ` "label.default-sell-price": "Default sell price",` -- it's just one of several `label.` entries in that file.

**Comment by CarlosNZ** [comment]
Weird, I tried it on a different machine with a freshly-cloned repo and I'm still getting the same error.

Running on macOS 12.3.

I cloned this repo and ran the your tests and got similar errors:
<img width="642" alt="Screen Shot 2022-04-05 at 9 34 35 AM" src="https://user-images.githubusercontent.com/5456533/161636331-d22a1c7a-7c79-48fa-9faa-711d48c21eea.png">

**Comment by mxmvshnvsk** [comment]
The current tests do not cover the codebase. I'll deal with this issue later. I'll try run clean installation on another machine later. I tried run in debug mode at this point on my dev machine. Everything ok. Did u try run debug or log on your current environment?

**Comment by mxmvshnvsk** [comment]
The current tests do not cover the codebase. I'll deal with this issue later. I'll try run clean installation on another machine later.

**Comment by mxmvshnvsk** [comment]
@CarlosNZ, I think I found the problem =) I looked at your [update commit](https://github.com/openmsupply/openmsupply-client/commit/781342bcd7d5ae015b0506f21bdf8f28fed280b6), you update package version, but not turn on flat json option in [config](https://github.com/openmsupply/openmsupply-client/blob/i18n-unused/i18n-unused.config.js), please, add follow option (it's in nested mode by default):

```javascript
flatTranslations: true
```

**Comment by CarlosNZ** [comment]
> @CarlosNZ, I think I found the problem =) I looked at your [update commit](https://github.com/openmsupply/openmsupply-client/commit/781342bcd7d5ae015b0506f21bdf8f28fed280b6), you update package version, but not turn on flat json option in [config](https://github.com/openmsupply/openmsupply-client/blob/i18n-unused/i18n-unused.config.js), please, add follow option (it's in nested mode by default):

Ah right, yes, my bad -- I didn't realise we needed to manually add that config option. Working well now, thank you :) 

I'll close this issue now, but I just wanted to point out one tiny little thing you might want to address at some point: When running `mark` or `remove` your script removes a final line break in each file (even when there's nothing changed), which is different to how our auto-formatter (Prettier) does it (adds a final line break). So any file that we've manually saved shows up in the git diff after running this script, even if nothing has actually changed. 
<img width="434" alt="Screen Shot 2022-04-06 at 9 06 53 AM" src="https://user-images.githubusercontent.com/5456533/161849255-42d94e44-0d04-4eef-a5da-4df731330e6b.png">
